blocks/log.py

```diff
@@ -10,265 +9,61 @@
-    def get_default_value(self, key):
-        """Returns the default value set for the 'key'.
-
-        Returns
-        -------
-        The default value for the `key`, or ``None`` if not set.
-
-        """
-        pass
-
-    def add_record(self, time, key, value):
-        """Adds a record to the log.
-
-        If `value` equals to the default value for the `key`, nothing is
-        done.
-
-        """
-        self._check_time(time)
-        default_value = self.get_default_value(key)
-        if value != default_value:
-            self._add_record(time, key, value)
-
-    @abstractmethod
-    def _add_record(self, time, key, value):
-        """Adds a record to the log.
-
-        The implementation method to be overridden.
-
-        """
-        pass
-
-    def fetch_record(self, time, key):
-        """Fetches a record from the log.
-
-        If no such 'key' for the `time` is found or if the value for the
-        key is ``None``, the default value for the 'key' is returned.
-
-        """
-        self._check_time(time)
-        value = self._fetch_record(time, key)
-        if value is not None:
-            return value
-        return self.get_default_value(key)
-
-    @abstractmethod
-    def _fetch_record(self, time, key):
-        """Fetches a record from the log.
-
-        The implementation method to be overridden.
+    def __init__(self):
+        super(TrainingLog, self).__init__(dict)
+        self.status = {
+            'iterations_done': 0,
+            'epochs_done': 0,
+            '_epoch_ends': []
+        }
 
-        """
-        pass
+    def __reduce__(self):
+        constructor, args, _, _, items = super(TrainingLog, self).__reduce__()
+        return constructor, (), self.__dict__, _, items
 
     def __getitem__(self, time):
         self._check_time(time)
-        return TrainingLogRow(self, time)
+        return super(TrainingLog, self).__getitem__(time)
+
+    def _check_time(self, time):
+        if not isinstance(time, Integral) or time < 0:
+            raise ValueError("time must be a positive integer")
 
     @property
     def current_row(self):
-        return self[self.status.iterations_done]
+        return self[self.status['iterations_done']]
 
     @property
     def previous_row(self):
-        return self[self.status.iterations_done - 1]
+        return self[self.status['iterations_done'] - 1]
 
     @property
     def last_epoch_row(self):
-        return self[self.status._epoch_ends[-1]]
-
-    @abstractmethod
-    def __iter__(self):
-        """Returns an iterator over time-key-value triples of the log."""
-        pass
-
-    @abstractmethod
-    def get_row_iterator(self, time):
-        """Returns an iterator over key-value pairs of a row of the log."""
-        pass
-
-    def _check_time(self, time):
-        if not isinstance(time, int) or time < 0:
-            raise ValueError("time must be a positive integer")
+        return self[self.status['_epoch_ends'][-1]]
```

Comment:
In the other PR I replaced it with a `self.current_entry['epoch_ended'] = True` entry in the log, and that actually works just fine. You can just do `epoch_ends = [key for key, value in self.log.items() if 'epoch_ended' in value]`. Slightly longer than `epoch_ends = self.log.status['_epoch_ends']`, but a lot more consistent, since right now we're effectively saving log entries into the status for some reason.